packages/server/src/utils/workflow.ts

```diff
@@ -180,16 +183,44 @@ function transformWorkflowEvent(
   if (event instanceof AgentToolCallResult) {
     const rawOutput = event.data.raw;
 
+    const isValidToolOutput = rawOutput && typeof rawOutput === "object";
+    if (!isValidToolOutput) return event; // skip transform if valid tool output is not found
+
     // if AgentToolCallResult contains sourceNodes, convert it to SourceEvent
     if (
-      rawOutput &&
-      typeof rawOutput === "object" &&
       "sourceNodes" in rawOutput // TODO: better use Zod to validate and extract sourceNodes from toolCallResult
     ) {
       return toSourceEvent(
         rawOutput.sourceNodes as unknown as NodeWithScore<Metadata>[],
       );
     }
+
+    // if AgentToolCallResult contains artifact, it's output from codeGenerator tool, convert it to ArtifactEvent with code data
+    if (
+      "artifact" in rawOutput // TODO: better use Zod to validate and extract artifact from toolCallResult
+    ) {
+      const { file_path, code } = rawOutput.artifact as {
+        file_path: string;
+        code: string;
+      };
+
+      const filename = file_path.split("/").pop();
+      const language = filename?.split(".").pop();
+
+      // TODO: make it able to streaming code and document content?
```

Comment:
yes, we should be able to stream it - can be done in later PR - probably using streamed tool calls? see https://github.com/run-llama/LlamaIndexTS/issues/1840